### ISSUE TITLE
[Feature Request] CSI Spec Volume Creation

### ISSUE THREAD

**Issue by sbouts** [hashicorp/nomad#8212]
Currently, HashiCorp Nomad implements a subset of the CSI spec found [here](https://github.com/container-storage-interface/spec/blob/master/spec.md). This feature request is for the implementation of the volume related methods. 

At the bottom of this feature request I have included an overview of the CSI spec methods as well as the ones Nomad has currently implemented.

## Usecase
Ceph is an Open Source scalable distributed storage system that is widely adopted among enterprises and strongly liked for its scalability and performance.

Implementing the missing CSI spec methods  would enable the use of the [ceph-csi](https://github.com/ceph/ceph-csi) driver in Nomad to dynamically provision persistent volumes hosted on [Ceph](https://ceph.io/) for Nomad workloads.

This would bring Nomad on par with other CO like Kubernetes and Mesos.

## The Problem
There have been attempts by the community to try and get the ceph-csi <==> Nomad communication to work. 
* [On the forum](https://discuss.hashicorp.com/t/help-with-csi-driver-ceph-and-configmaps/7933)
* [In issues on Github](https://github.com/hashicorp/nomad/issues/7668#issuecomment-629852292)

The problem boils down to Nomad not having fully implemented the CSI spec methods for all the [Controller capabilities](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllergetcapabilities) and [Node capabilities](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetcapabilities) the ceph-csi driver (and others) offers. The ceph-csi driver expects volumes to be dynamically provisioned and mounted on the fly (see [here](https://github.com/ceph/ceph-csi#ceph-csi-features-and-available-versions)) and not being created beforehand and explicitly put into e.g. the Nomad Volume Specification.

### Ceph-csi Capabilities
The ceph-csi driver currently implements the following CSI Spec capabilities:

**Note:** The CSI methods for RBD and CephFS may differ in implementation, but adhere to the same CSI specification.

| Ceph component | CSI Type | CSI Capabilities | Related CSI Methods | Implemented by Nomad?
|-------- |-------------- |------ | ------- | -------
| RBD  | Controller  | CREATE_DELETE_VOLUME  | [CreateVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume)  | :x: |
|   |   |  | [DeleteVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume)  | :x: |
| RBD  | Controller  | CREATE_DELETE_SNAPSHOT  | [CreateSnapshot](https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot)  | :x: |
|   |   |  | [DeleteSnapshot](https://github.com/container-storage-interface/spec/blob/master/spec.md#deletesnapshot)  | :x: |
| RBD  | Controller  | CLONE_VOLUME  | [CreateVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume)  | :x: |
| RBD  | Controller  | EXPAND_VOLUME  | [ControllerExpandVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume)  | :x: |
| RBD  | Node  | STAGE_UNSTAGE_VOLUME  | [NodeStageVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L499)  | :heavy_check_mark: |
|   |   |   | [NodeUnstageVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L547)  | :heavy_check_mark: |
| RBD  | Node  | GET_VOLUME_STATS  | [NodeGetVolumeStats](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetvolumestats)  | :x: |
| RBD  | Node  | EXPAND_VOLUME  | [NodeExpandVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodeexpandvolume)  | :x: |

| Ceph component | CSI Type | CSI Capabilities | Related CSI Methods | Implemented by Nomad?
|-------- |-------------- |------ | ------- | -------
| CephFS  | Controller  | CREATE_DELETE_VOLUME  | [CreateVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume)  | :x: |
|   |   |  | [DeleteVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume)  | :x: |
| CephFS  | Controller  | EXPAND_VOLUME  | [ControllerExpandVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume)  | :x: |
| CephFS  | Node  | STAGE_UNSTAGE_VOLUME  |  [NodeStageVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L499)  | :heavy_check_mark: |
|   |   |   | [NodeUnstageVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L547)  | :heavy_check_mark: |
| CephFS  | Node  | GET_VOLUME_STATS  |  [NodeGetVolumeStats](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetvolumestats)  | :x: |

## The Fix
The easier part of the fix is the actual (dummy) implementation of the missing CSI methods in the [Nomad client](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go). 

The harder part of the fix is how to implement a logical flow in e.g. the [Job Stanza](https://www.nomadproject.io/docs/job-specification/job/) and the [Volume Specification](https://www.nomadproject.io/docs/commands/volume/register/#volume-specification)  to service both the creation and mounting of NEW volumes through the [CreateVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume) API (like ceph-csi expects it) as well as the mounting of EXISTING volumes (like [this Nomad example](https://www.hashicorp.com/blog/hashicorp-nomad-container-storage-interface-csi-beta/)).

Of course I would be more than willing to give a shot at helping to implement the missing methods. Getting the flow right from HCL to the mounting of volumes in Nomad workloads ties more into the core of Nomad and I feel that should be left up to you guys to implement.

## CSI Spec Method Overview And Current Nomad Adoption

| Identity Method   | Implemented?  | Nomad Link  |
|-------- |-------------- |------ |
| GetPluginInfo |    :heavy_check_mark:  | [GetPluginInfo](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L194) |
| GetPluginCapabilities |  :heavy_check_mark:  | [GetPluginCapabilities](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L216) |
| Probe | :heavy_check_mark: | [Probe](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L172) |

| Controller Method | Implemented? | Nomad Link | CSI Spec Link
|-------- |-------------- |------ | ----------
| CreateVolume | :x: | | [CreateVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#createvolume) |
| DeleteVolume | :x: | | [DeleteVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#deletevolume) |
| ControllerPublishVolume |  :heavy_check_mark:  | [ControllerPublishVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L258) |
| ControllerUnpublishVolume |  :heavy_check_mark:  | [ControllerUnpublishVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L300) |
| ValidateVolumeCapabilities | :heavy_check_mark: | [ValidateVolumeCapabilities](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L329) |
| ListVolumes | :x: | | [ListVolumes](https://github.com/container-storage-interface/spec/blob/master/spec.md#listvolumes) |
| GetCapacity | :x: | | [GetCapacity](https://github.com/container-storage-interface/spec/blob/master/spec.md#getcapacity) |
| ControllerGetCapabilities |  :heavy_check_mark:  | [ControllerGetCapabilities](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L239) |
| CreateSnapshot | :x: | | [CreateSnapshot](https://github.com/container-storage-interface/spec/blob/master/spec.md#createsnapshot) |
| DeleteSnapshot | :x: | | [DeleteSnapshot](https://github.com/container-storage-interface/spec/blob/master/spec.md#deletesnapshot) |
| ListSnapshots | :x: | | [ListSnapshots](https://github.com/container-storage-interface/spec/blob/master/spec.md#listsnapshots) |
| ControllerExpandVolume | :x: | | [ControllerExpandVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume) |
| ControllerGetVolume | :x: | | [ControllerGetVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#controllerexpandvolume) |

| Node Method | Implemented? | Nomad Link | CSI Spec Link
|-------- |-------------- |------ | ------- 
| NodeStageVolume |  :heavy_check_mark:  | [NodeStageVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L499) |
| NodeUnstageVolume |  :heavy_check_mark:  | [NodeUnstageVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L547) |
| NodePublishVolume |  :heavy_check_mark:  | [NodePublishVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L584) |
| NodeUnpublishVolume |  :heavy_check_mark:  | [NodeUnpublishVolume](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L618) |
| NodeGetVolumeStats | :x: | | [NodeGetVolumeStats](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetvolumestats) |
| NodeExpandVolume | :x: | | [NodeExpandVolume](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodeexpandvolume) |
| NodeGetCapabilities |  :heavy_check_mark: | [NodeGetCapabilities](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L450) |
| NodeGetInfo |  :heavy_check_mark:  | [NodeGetInfo](https://github.com/hashicorp/nomad/blob/master/plugins/csi/client.go#L468) |

**Comment by tgross** [comment]
Hi @sbouts, thanks for opening this issue! Yes as you've noted we didn't implement the volume creation workflow. We had this slated for a potential Phase 2 of work but I don't think we realized this was a blocker for using plugins like Ceph.

As you've noted here, there's a good bit of work required to make this happen, so I'm going to tag my colleagues @galeep and @yishan-lin to make sure it's on their radar as some of our planning gets solidified.

**Comment by ryanmickler** [comment]
This feature request from on the terraform nomad provider is also relevant for this work.
https://github.com/terraform-providers/terraform-provider-nomad/issues/102

**Comment by tgross** [comment]
Thanks for the link @ryanmickler. That feature request is to cover the existing Nomad registration APIs, not the CSI APIs being discussed here.

**Comment by ryanmickler** [comment]
Is there a seperate issue to implement the `parameters` block, not the entire volume creation spec? Currently, that's all thats limiting use of the `ceph-csi` plugin

see here:
https://github.com/hashicorp/nomad/issues/7668#issuecomment-645878741

**Comment by tgross** [comment]
@ryanmickler reading that comment and then digging into https://github.com/container-storage-interface/spec/issues/387 it doesn't look like this is part of the CSI spec yet for the RPCs we support, is it?

**Comment by ryanmickler** [comment]
A quick inspection looks like `NodeStageVolume` (mounts the volume to a staging path on the node.)
https://github.com/ceph/ceph-csi/blob/47d5b60af8d48574ff6d11ca37dbff5a6f56815b/internal/rbd/nodeserver.go#L116
is calling `genVolFromVolumeOptions` on line 171
https://github.com/ceph/ceph-csi/blob/47d5b60af8d48574ff6d11ca37dbff5a6f56815b/internal/rbd/nodeserver.go#L171
Then inside that function, we are hitting our `missing required parameter pool` error here:
https://github.com/ceph/ceph-csi/blob/be9e7cf956c378227ff43e0194410468919766b7/internal/rbd/rbd_util.go#L694

**Comment by tgross** [comment]
Support for those fields were added in https://github.com/hashicorp/nomad/pull/7957

**Comment by ryanmickler** [comment]
Sure, perhaps either my nomad version (11.3) is missing that PR, or those parameters arent getting passed to `NodeStageVolume` properly. Because '`pool = "<value>"` is set in my config.

`E0728 04:47:20.003465       1 utils.go:163] ID: 23 Req-ID: csi-test-0 GRPC error: rpc error: code = Internal desc = missing required parameter pool`

**Comment by tgross** [comment]
@ryanmickler we should dig into this but let's not clutter up this feature request issue with debugging that. Can you open a new issue and include your volume spec and any relevant logs?

**Comment by tgross** [comment]
Just an update on this long-awaited feature, I'll be working on this over the next few weeks.

**Comment by travisghansen** [comment]
As a csi driver author I had someone use cli tools to create the volumes etc and then copy responses into nomad to work that way. It’s not ideal but functional. Great to hear this is being worked on and if any questions arise happy to collaborate with the experience I have from the other side of the equation.

- https://github.com/democratic-csi/democratic-csi/blob/master/docs/nomad.md

**Comment by tgross** [comment]
Just a heads up that I've split ExpandVolume, GetCapacity, and NodeGetVolumeStatus out to their own issues #10324 and #10325, because they're likely to miss 1.1.0 and have to get shipped in a patch version shortly thereafter. Once the create/snapshot E2E testing is done (#10322) and I've updated the API docs, this issue will be closed and shipped for Nomad 1.1.0 (public beta a few weeks out yet).
